core/src/main/java/org/apache/iceberg/SchemaUpdate.java

```diff
@@ -340,26 +407,30 @@ public Type field(Types.NestedField field, Type fieldResult) {
     }
 
     @Override
-    public Type list(Types.ListType list, Type result) {
+    public Type list(Types.ListType list, Type elementResult) {
       // use field to apply updates
-      Type elementResult = field(list.fields().get(0), result);
-      if (elementResult == null) {
+      Types.NestedField elementField = list.fields().get(0);
+      Type elementType = field(elementField, elementResult);
+      if (elementType == null) {
         throw new IllegalArgumentException("Cannot delete element type from list: " + list);
       }
 
-      if (list.elementType() == elementResult) {
+      Types.NestedField elementUpdate = updates.get(elementField.fieldId());
+      boolean isElementOptional = elementUpdate != null ? elementUpdate.isOptional() : list.isElementOptional();
+
+      if (isElementOptional == elementField.isOptional() && list.elementType() == elementType) {
```

Comment:
should we also check if the field's doc is updated?

Reply:
Because elements, keys, and values are fields, they have doc strings, but these aren't actually exposed in the type API because these are accessed as types instead of fields. For example, `list.elementType()`.

There is no direct getter, like `list.elementDoc()`, and I don't think it would be useful to add those, since you can document the list's elements as part of the list. Does that make sense?

Reply:
Yes, it does!